### DISCUSSION TITLE
Putting a string together from ints

### DISCUSSION THREAD

**Discussion by rvjansen** [adesutherland/CREXX#69] [Progress Updates]
I have this, a working length() in rxas around strlen:
```
/*
 * prototype length() function
 * no error checking
 */
.globals=0

length()  .locals=4 .expose=global.length
          strlen r2,a1  /* first argument          */
          ret    r2     /* return length of string */

```
Substr() needs length(), so I am at this point:
```
/*
 * rexx SUBSTR bif in rxas. Uses length.rxas
 * rudimentary error checking
 */

          .globals=0
length()  .expose=global.length
substr()  .locals=14 .expose=global.substr
          /* string    = arg1    */
          /* n (start) = arg2    */
          ilt r1,a0,2 /* less than 2 arguments */
          brt wrongargs,r1
          igt r1,a0,4 /* more than 4 arguments */
          brt wrongargs,r1
          ieq r1,a0,2
          brt twoargs,r1 /* we have two arguments */
wrongargs:
          /* we are here if there are not enough or too many arguments */
          say "substr needs at least two and at most 4 arguments"
          exit
twoargs:
          /* we want to know the lenght of the passed string */
          load r3,1 /* there is one argument for the call to length() */
          move r4,a1 /* and it is the string in a1 */
          call r10,length(),r3 /* what is the length of the string arg */
          /* now loop characters from a1 from the start pos in a2 to (length) r1 */
loop1:
          /* a2:int argument 2, start position */
          /* r10:int length of arg1, end position */
          /* r4:string argument1, string */
          igt r11,a2,r10
          brt endloop1,r11
          strchar r13,r4,a2
          itos a2
          say a2
          itos r13
          say r13
          itos r13
          say r13
          iadd a2,a2,1
          br  loop1
endloop1:
          say "in endloop1"
          exit

```
but now I am at the point of putting the new string together, and I don't know how to do that from integers.
(had some correspondence with Peter, and he sent me something)

Also notice that I had, for a call within a call, to move a1 to r4, because the inner call clobbered that (I think). Is that how it is supposed to work?

**Comment by adesutherland** [comment]
Firstly I am amazed - I didn't really expect people to code in REXX Assembler!!

The move from a1 to r4 is needed, I think, because you have to marshal the arguments after r3 (in this case). The good news is that move just moves the pointer and does not copy data/strings or anything like like - fast. The bad is after "move r4,a1" a1 is no longer a valid register and cannot be used (this might be a blunder in the design!!)

We have instruction "appendchar" -> "Append Concat Char op2 (as int) on op1". This might be what you need?

**Comment by rvjansen** [comment]
Will continue with appendchar - missed that.
Well, at least move is not a copy like mvc and mvcl in s/360 ;-)

René.

**Comment by rvjansen** [comment]
what am I doing wrong here?

```
/* test appendchar */
options levelb

say 'test appendchar'
stringout=''
assembler
do
  appendchar stringout,55
  say stringout
end
```
yields:
rxdas testappendchar
```
* CONSTANT POOL - Size 0x78

* 0x000000 PROC   main() @ 0x000000 (locals=4)
* 0x000040 STRING "test appendchar"

* CODE SEGMENT - Size 0x3

.globals=0

main()          .locals=4
                say "test appendchar"                        * 0x000000 Say op1
                exit                                         * 0x000002 Exit

```

and the test is because when I use append char in my substr source, it seems to put string representations of the int in the appended string, like:
```
r13:
53
r14:
5
r13:
54
r14:
56
r13:
55
r14:
567
r13:
56
r14:
5678
in endloop1
5678
```

**Reply by Peter-Jacob** [reply]
@rvjansen: I think your proposal clearly shows the strength of REXX assembler. Performance-wise I think we should divide the code into 2 portions: 
1. setup and validation of input parameters
   keep this in assembler, it's easier to handle than put it in the C-Code 
2. creating the substring
   this should be a new instruction to avoid the call of too many instructions.

my two pennyworths!

**Comment by rvjansen** [comment]
on the other hand, this clearly works:
```
.globals=0

main()  .locals=4
          load r1,"beginofstring"
          say r1
          load r2,65
          appendchar r1,r2
          say r1
          exit
```
beginofstring
beginofstringA

**Reply by rvjansen** [reply]
ok, before I call it a day:
```
.globals=0

main()  .locals=11
          load r1,"beginofstring"
          say r1
          load r2,65
          load r3,66
          appendchar r1,r2
          appendchar r1,r3    
          say r1
          load r5,1
          load r6,2
          load r7,3
          strchar r8,r1,r5
          itos r8
          say r8
          strchar r9,r1,r6
          itos r9
          say r9
          strchar r10,r1,r7
          itos r10
          say r10
          appendchar r1,r10
          say r1
          exit
```
gives:
```
beginofstring
beginofstringAB
50
53
55
beginofstringAB7
```
but I was expecting beginofstringABg
?

**Comment by adesutherland** [comment]
I think this culprit is this which I didn't re-visit when I did UTF. Also the "-'0'" explains the mismatch. Anyway needs fixing ...


        START_INSTRUCTION(STRCHAR_REG_REG_REG) CALC_DISPATCH(3);
            DEBUG("TRACE - STRCHAR R%d,R%d,R%d\n", (int)REG_IDX(1), (int)REG_IDX(2), (int)REG_IDX(3));
            v2 = op2R;
            v3 = op3R;
            i1 = v2->string_value[v3->int_value - 1] - '0';
            REG_RETURN_INT(i1);
            DISPATCH;


But all is not lost! We have the following (and as a hidden bit of undocumentedness - each string has a current position)

 * SETSTRPOS_REG_REG - Set String (op1) charpos set to op2
 * GETSTRPOS_REG_REG - Get String (op2) charpos into op1
 * STRCHAR_REG_REG - op1 (as int) = op2[charpos]
```   
      /* r0 is the string */
      load r1,40 /* I want the 40th character */
      setstrpos r0,r1
      strchar r2,r0
```
The instructions are optimised so that if you now want to get the 41st character it knows just to move forward 1 character and not 
search through the whole string from the beginning again.


The broken STRCHAR_REG_REG_REG needs to be redone to combine SETSTRPOS_REG_REG and STRCHAR_REG_REG in one go.

**Reply by Peter-Jacob** [reply]
the conversion from string to integer using **string_value- '0'** works of course only for ASCII/EBCDIC, I didn't have UTF8 in mind. I'll look at Adrian's proposal using **_SETSTRPOS_REG_REG_ and STRCHAR_REG_REG**, but I am still shy with UTF8

**Reply by Peter-Jacob** [reply]
With a little help from my friends: the problem is fixed!
beginofstring
beginofstringAB
98
101
103
beginofstringABg

**Reply by rvjansen** [reply]
very cool, thank you!

**Comment by rvjansen** [comment]
Hi Peter,

Yes, I think so too. There were a few reasons I started this:

- wanted to experiment with rxas and learn the linkage conventions between rxas modules
- wanted to test the things that are already there (and found some)
- as a prelude to writing them in Rexx levelb, and checking the difference with generated rxas
- having a good reference for performance differences between Rexx, rxas, C and handcrafted assembler

I still see some onion-skin model for calling code - when there is no C implementation available, we should be able to call a Rexx implementation; in some cases, handwritten assembler will be fastest, but might not be available for all instruction set architectures. The big difference is native linkage and Rexx-type linkage.

A secondary goal is to keep the kernel very small; ideally, its should be bootstrappable at some point. If I, for example wanted to have a JVM implementation of the interpreter (not that imaginary, because then I could have compatibility for CMS user-written Rexx Pipelines in NetRexx Pipelines) I would like to be able to have all the BIF’s in a Rexx version, with Rexx levelb linkage. Not that I would not use the C-versions if they are much faster - but I like to measure things. Another thought is that not all compilers for embedded environments would have the same strength or feature set.

I am sure we will discuss this a lot. For the moment, we should plan well and try not to duplicate efforts (and making a C version of something that can be Rexx or (native or Rexx) assembler does not count as duplication. For speed, I used the already available STRLEN in the LENGHT() bif, but I plan to measure the difference when I have rxas or Rexx levelb versions.

About the validation of input: this should also be well-placed. In compiled code, the compiler can do it; in the interpreter VM, we should choose where to do local validation in functions, or when to use generic validation. The examples from the standard have them centralised, but seem to do quite some validation that is for some functions only.

I will leave the design of the linkage mechanism to another post.

Best regards,

René.
